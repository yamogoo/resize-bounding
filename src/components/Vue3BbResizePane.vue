<template>
  <div
    :class="[`${prefix}__pane`, `--${direction}`]"
    data-testid="bb-resize-pane"
    :style="[styles?.pane, paneComputedStyle]"
  >
    <div
      ref="refPane"
      data-testid="bb-resize-splitter"
      :class="[`${prefix}__splitter`, isFocused ? 'show' : 'hide']"
      :style="[styles?.splitter, splitterComputedStyle]"
    >
      <div
        v-if="isFocused || options?.knob?.constantlyShow"
        data-testid="bb-resize-knob"
        :class="[`${prefix}__knob`]"
        :style="[styles?.knob, knobComputedStyle]"
      ></div>
    </div>
  </div>
</template>

<script setup lang="ts">
import {
  onMounted,
  onUnmounted,
  ref,
  type Ref,
  computed,
  type ComputedRef,
  type HTMLAttributes,
} from "vue";

import { BBResize } from "./typings";

const props = withDefaults(defineProps<Props>(), {
  direction: PaneDirections.RIGHT,
});

const emits = defineEmits<{
  (e: Emits.DRAG_START, data: PaneEmittedData): void;
  (e: Emits.DRAG_MOVE, data: PaneEmittedData): void;
  (e: Emits.DRAG_END, data: PaneEmittedData): void;
  (e: Emits.FOCUS, isFocused: boolean): void;
}>();

const refPane: Ref<HTMLDivElement | null> = ref(null);

const paneComputedStyle = computed(() => {
  const _width = props.options?.width ?? 4;

  if (refPane.value && _width) {
    const _styles = paneBaseStyles(_width);
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
});

const splitterComputedStyle = computed(() => {
  console.log(props.options);
  const _width = props.options?.width;

  if (refPane.value && _width) {
    const _styles = splitterBaseStyles(
      _width,
      props.options?.position ?? "center",
    );
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
});

const isFocused = ref(false);
let isResizing = false;

const knobComputedStyle = computed(() => {
  const isHorizontal = checkIsHorizontal();
  return { transform: `rotate(${isHorizontal ? 0 : 90}deg)` };
});

const checkIsHorizontal = (): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(props.direction);
const checkIsVertical = (): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(props.direction);

const updateCursor = (state: boolean) => {
  let cursorActive: string;

  if (checkIsHorizontal()) {
    cursorActive = props.options?.cursor?.horizontal ?? "col-resize";
  } else if (checkIsVertical()) {
    cursorActive = props.options?.cursor?.vertical ?? "row-resize";
  } else cursorActive = "auto";

  if (refPane.value) refPane.value.style.cursor = state ? cursorActive : "auto";
};

const emitFocus = (state: boolean) => {
  emits(Emits.FOCUS, state);
};

const onFocus = (e: PointerEvent, isFocused: boolean): void => {
  e.stopPropagation();
  onSelected(isFocused);
};

const onSelected = (state: boolean): void => {
  isFocused.value = state;
  updateCursor(state);
  emitFocus(state);
};

const onDragStart = (e: PointerEvent): void => {
  e.preventDefault();
  e.stopImmediatePropagation();

  isResizing = true;
  onSelected(isResizing);

  const el = e.currentTarget as HTMLDivElement;
  el.setPointerCapture(e.pointerId);

  emits(Emits.DRAG_START, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });

  const onDragMove = (e: PointerEvent): void => {
    emits(Emits.DRAG_MOVE, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  const onDragEnd = (e: PointerEvent): void => {
    isResizing = false;

    onSelected(isResizing);
    updateCursor(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.releasePointerCapture(e.pointerId);

    el.removeEventListener("pointermove", onDragMove);
    el.removeEventListener("pointerup", onDragEnd);

    emits(Emits.DRAG_END, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  el.addEventListener("pointermove", onDragMove);
  el.addEventListener("pointerup", onDragEnd);
};

const onDragCancel = (e: PointerEvent): void => {
  const el = e.currentTarget as HTMLDivElement;

  isResizing = false;
  el.releasePointerCapture(e.pointerId);

  emits(Emits.DRAG_END, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });
};

const addEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.addEventListener("pointerenter", (e) => onFocus(e, true));
    el.addEventListener("pointerleave", (e) => onFocus(e, false));

    el.addEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

const removeEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.removeEventListener("pointerenter", (e) => onFocus(e, true));
    el.removeEventListener("pointerleave", (e) => onFocus(e, false));

    el.removeEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

onMounted(addEventListeners);
onUnmounted(removeEventListeners);

defineExpose({ refPane });
</script>

<script lang="ts">
export interface Props {
  prefix: string;
  direction?: PaneDirectionKey;
  options?: Partial<BBResize.PaneOptions>;
  styles?: Partial<BBResize.Styles>;
}

export enum PaneDirectionAliases {
  HORIZONTAL = "h",
  VERTICAL = "v",
}

export enum PaneDirections {
  LEFT = "l",
  RIGHT = "r",
  BOTTOM = "b",
  TOP = "t",
}

export type PaneDirectionKey = PaneDirections | PaneDirections | string;

export enum Emits {
  FOCUS = "focus",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

export const paneBaseStyles = (
  size: number,
): Record<PaneDirections, HTMLAttributes["style"]> => {
  const _offset = `${size / 2}px`;

  return {
    l: { top: "0px", left: _offset, width: "0px", height: "100%" },
    r: { top: "0px", right: _offset, width: "0px", height: "100%" },
    t: { left: "0px", top: _offset, width: "100%", height: "0px" },
    b: { left: "0px", bottom: _offset, width: "100%", height: "0px" },
  };
};

export const splitterBaseStyles = (
  size: number,
  position: BBResize.PanePosition,
): Record<PaneDirections, HTMLAttributes["style"]> => {
  const _size = `${size}px`;

  let _offset: string = BBResize.SplitterPositions.CENTER;

  switch (position) {
    case BBResize.SplitterPositions.CENTER:
      _offset = `0px`;
      break;
    case BBResize.SplitterPositions.EXTERNAL:
      _offset = `${size / 2}px`;
      break;
    case BBResize.SplitterPositions.INTERNAL:
      _offset = `-${size / 2}px`;
      break;
  }

  return {
    l: { right: _offset, width: _size, height: "100%" },
    r: { left: _offset, width: _size, height: "100%" },
    t: { bottom: _offset, height: _size, width: "100%" },
    b: { top: _offset, height: _size, width: "100%" },
  };
};
</script>

<style lang="scss">
$__show-color: blue;
$__pressed-color: #3655e171;

.bb-resize {
  &__pane {
    position: absolute;
    display: block;
    z-index: 9999;

    // &.--l {

    //   .bb-resize__pane__splitter {
    //   }
    // }

    // &.--r {

    //   .bb-resize__pane__splitter {
    //   }
    // }

    // &.--b {

    //   .bb-resize__pane__splitter {
    //   }
    // }

    // &.--t {

    //   .bb-resize__pane__splitter {
    //   }
    // }
  }

  &__splitter {
    display: flex;
    align-items: center;
    justify-content: center;
    z-index: 9999;
  }

  &__splitter {
    position: absolute;

    &.show {
      background-color: $__show-color;
    }

    &.pressed {
      background-color: $__pressed-color;
    }
  }

  &__knob {
    position: absolute;
    width: 12px;
    height: 120px;
    border-radius: 12px;
    background-color: red;
    margin: auto;
  }
}
</style>
