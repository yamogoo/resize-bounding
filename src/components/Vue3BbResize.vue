<template>
  <div
    v-bind="$attrs"
    ref="refRoot"
    data-testid="bb-resize-container"
    :class="options.prefix"
    :style="[computedStyle, styles?.container]"
  >
    <slot :width :height></slot>
    <template v-for="({ show, direction }, idx) in panes" :key="idx">
      <Vue3BbResizePane
        v-if="!disabled && show"
        :prefix="options.prefix ?? ''"
        :direction="direction"
        :options="options.pane"
        :styles
        @focus="
          (isFocused) => {
            $emit(Emits.FOCUS, { state: isFocused, direction });
          }
        "
        @drag:start="onDragStart"
        @drag:move="onDragMove"
        @drag:end="onDragEnd"
      />
    </template>
  </div>
</template>

<script lang="ts">
import {
  ref,
  computed,
  defineComponent,
  type PropType,
  type ComputedRef,
  type HTMLAttributes,
} from "vue";

import { deepMerge } from "./utils";

import type { BBResize } from "./typings";

import Vue3BbResizePane, {
  PaneDirectionAliases,
  PaneDirections,
  type PaneDirectionKey,
  type PaneEmittedData,
} from "./Vue3BbResizePane.vue";

export enum Emits {
  FOCUS = "focus",
  UPDATE_WIDTH = "update:width",
  UPDATE_HEIGHT = "update:height",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

const defaultOptions: BBResize.Options = {
  prefix: "bb-resize",
  pane: {
    width: 12,
    position: "internal",
    knob: {
      constantlyShow: true,
    },
  },
};

export default defineComponent({
  setup(props, { emit }) {
    const refRoot = ref<HTMLDivElement | null>(null);

    let { width: newWidth, height: newHeight } = props;
    let prevWidth = newWidth,
      prevHeight = newHeight;

    const computedStyle: ComputedRef<HTMLAttributes["style"]> = computed(() => {
      const _width = {
        width: `${props.width}px`,
        minWidth: `${props.width}px`,
      };

      const _height = {
        height: `${props.height}px`,
        minHeight: `${props.height}px`,
      };

      return {
        position: "relative",
        ...(props.width && _width),
        ...(props.height && _height),
      };
    });

    const panes = computed(() => ({
      left: {
        show:
          RegExp(PaneDirections.LEFT).test(props.directions) ||
          RegExp(PaneDirectionAliases.HORIZONTAL).test(props.directions),
        direction: PaneDirections.LEFT,
      },
      right: {
        show:
          RegExp(PaneDirections.RIGHT).test(props.directions) ||
          RegExp(PaneDirectionAliases.HORIZONTAL).test(props.directions),
        direction: PaneDirections.RIGHT,
      },
      bottom: {
        show:
          RegExp(PaneDirections.BOTTOM).test(props.directions) ||
          RegExp(PaneDirectionAliases.VERTICAL).test(props.directions),
        direction: PaneDirections.BOTTOM,
      },
      top: {
        show:
          RegExp(PaneDirections.TOP).test(props.directions) ||
          RegExp(PaneDirectionAliases.VERTICAL).test(props.directions),
        direction: PaneDirections.TOP,
      },
    }));

    let startWidth = props.width ?? 0,
      startHeight = props.height ?? 0;

    let startX = 0,
      startY = 0;

    const options = computed(() => {
      return deepMerge(defaultOptions, props.options);
    });

    const onDragStart = ({ x, y, dir }: PaneEmittedData): void => {
      startWidth = props.width ?? 0;
      startHeight = props.height ?? 0;

      startX = x;
      startY = y;

      emit(Emits.DRAG_START, dir);
    };

    const truncateInRange = (
      min: number,
      max: number | undefined,
      next: number,
    ): number => {
      const _max = max ?? Number.POSITIVE_INFINITY;

      if (next <= min) return min;
      else if (next >= _max) return _max;

      return next;
    };

    const onDragMove = ({ x, y, dir }: PaneEmittedData): void => {
      if (!refRoot.value) return;

      emit(Emits.DRAG_MOVE, dir);

      if (dir === PaneDirections.LEFT) {
        newWidth = startWidth + (startX - x);

        if (prevWidth === newWidth) return;

        const truncated = truncateInRange(
          props.minWidth,
          props.maxWidth,
          newWidth,
        );

        emit(Emits.UPDATE_WIDTH, truncated);
        prevWidth = truncated;
      } else if (dir === PaneDirections.RIGHT) {
        newWidth = startWidth + (x - startX);

        if (prevWidth === newWidth) return;

        const truncated = truncateInRange(
          props.minWidth,
          props.maxWidth,
          newWidth,
        );

        emit(Emits.UPDATE_WIDTH, truncated);
        prevWidth = truncated;
      } else if (dir === PaneDirections.TOP) {
        newHeight = startHeight + (startY - y);

        if (prevHeight === newHeight) return;

        const truncated = truncateInRange(
          props.minHeight,
          props.maxHeight,
          newHeight,
        );

        emit(Emits.UPDATE_HEIGHT, truncated);
        prevHeight = truncated;
      } else if (dir === PaneDirections.BOTTOM) {
        newHeight = startHeight + (y - startY);

        if (prevHeight === newHeight) return;

        const truncated = truncateInRange(
          props.minHeight,
          props.maxHeight,
          newHeight,
        );

        emit(Emits.UPDATE_HEIGHT, truncated);
        prevHeight = truncated;
      }
    };

    const onDragEnd = ({ dir }: PaneEmittedData): void => {
      emit(Emits.DRAG_END, dir);
    };

    return {
      refRoot,
      onDragStart,
      onDragMove,
      onDragEnd,
      truncateInRange,
      computedStyle,
      options,
      panes,
      PaneDirections,
      Emits,
    };
  },
  inheritAttrs: true,
  emits: {
    [Emits.FOCUS]: (data: { state: boolean; direction: string }) =>
      typeof data === "object" &&
      "state" in data &&
      typeof data.state === "boolean" &&
      "direction" in data &&
      typeof data.direction === "string",
    [Emits.UPDATE_WIDTH]: (width: number) => typeof width === "number",
    [Emits.UPDATE_HEIGHT]: (height: number) => typeof height === "number",
    [Emits.DRAG_START]: (dir: string) => typeof dir === "string",
    [Emits.DRAG_MOVE]: (dir: string) => typeof dir === "string",
    [Emits.DRAG_END]: (dir: string) => typeof dir === "string",
  },
  props: {
    width: {
      type: Number,
      default: undefined,
    },
    height: {
      type: Number,
      default: undefined,
    },
    minWidth: {
      type: Number,
      default: 0,
    },
    maxWidth: {
      type: Number,
      default: undefined,
    },
    minHeight: {
      type: Number,
      default: 0,
    },
    maxHeight: {
      type: Number,
      default: undefined,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    directions: {
      type: String as PropType<PaneDirectionKey | string>,
      default: `${PaneDirectionAliases.HORIZONTAL}${PaneDirectionAliases.VERTICAL}`,
    },
    alwaysShowKnob: {
      type: Boolean,
      default: false,
    },
    options: {
      type: Object as PropType<Partial<BBResize.Options>>,
      default: {},
    },
    styles: {
      type: Object as PropType<Partial<BBResize.Styles>>,
    },
  },
  name: "Vue3BbResize",
  components: { Vue3BbResizePane },
});
</script>
