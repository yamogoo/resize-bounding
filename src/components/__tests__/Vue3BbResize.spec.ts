import { describe, test, expect } from "vitest";
import { mount, shallowMount, VueWrapper } from "@vue/test-utils";

import { DataTestIds } from "./setup";

import Vue3BbResize, { Emits } from "@/components/Vue3BbResize.vue";
import {
  PaneDirectionAliases,
  PaneDirections,
  type PaneEmittedData,
} from "@/components/Vue3BbResizePane.vue";

import Vue3BbResizePane from "@/components/Vue3BbResizePane.vue";
import type { HTMLAttributes } from "vue";

const DEFAULT_PREFIX = "bb-resize";

describe("Vue3BbResize", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  describe("components", () => {
    test.each([
      ...Object.values(PaneDirections),
      ...Object.values(PaneDirectionAliases),
    ])(
      `Should not display any selected borders when set to "disabled" (--%s)`,
      (direction) => {
        const wrapper = mount(Vue3BbResize, {
          props: {
            directions: direction,
            disabled: true,
          },
        });

        const panEl = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);

        expect(panEl.length).toBe(0);
        expect(panEl.length).toMatchSnapshot();
      },
    );

    test.each(["rl", "ltrb", "bltr", "bl", "tr"])(
      "should render %s panes (panes computed value)",
      (directions) => {
        const wrapper = mount(Vue3BbResize, {
          props: {
            directions,
          },
        });

        const panEls = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);
        expect(panEls.length).toBe(directions.length);
        expect(panEls.length).toMatchSnapshot();
      },
    );

    test.each(["sd", "wo", ""])(
      "should not render any pane element",
      (directions) => {
        const wrapper = mount(Vue3BbResize, {
          props: {
            directions,
          },
        });

        const panEls = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);
        expect(panEls.length).toBe(0);
        expect(panEls.length).toMatchSnapshot();
      },
    );
  });

  describe("slots", () => {
    test.each(["<p>Inner Container</p>"])(
      "should render default slot",
      (slot) => {
        const wrapper = mount(Vue3BbResize, {
          props: {
            directions: "",
          },
          slots: {
            default: slot,
          },
        });

        const html = wrapper.find(`[data-testid="${DataTestIds.ROOT}"]`).html();
        expect(html).toContain(slot);
        expect(html).toMatchSnapshot();
      },
    );
  });

  describe("classes", () => {
    describe("prefix", () => {
      test.each(["my-prefix"])("should have custom prefix (%s)", (prefix) => {
        const wrapper = shallowMount(Vue3BbResize, {
          props: {
            options: {
              prefix,
            },
          },
        });

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.ROOT}"]`);
        const classes = rootEl.classes();

        expect(classes.toString()).toContain(prefix);
        expect(classes).toMatchSnapshot();
      });

      test("should have default prefix (%s)", () => {
        const wrapper = shallowMount(Vue3BbResize);

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.ROOT}"]`);
        const classes = rootEl.classes();

        expect(classes.toString()).toContain(DEFAULT_PREFIX);
        expect(classes).toMatchSnapshot();
      });
    });

    test.each([Object.values(PaneDirections)])(
      'should have "--%s" class',
      (direction) => {
        const wrapper = mount(Vue3BbResize, {
          props: {
            directions: direction,
          },
        });

        const panEl = wrapper.findAll(`[data-testid="${DataTestIds.PANE}"]`);

        expect(panEl[0].classes(`--${direction}`)).toBeTruthy();
        expect(panEl[0].classes(`--${direction}`)).toMatchSnapshot();
      },
    );
  });

  describe("events", () => {
    describe("drag", () => {
      const checkData = (
        eventName: string,
        wrapper: VueWrapper,
        direction: string,
      ): PaneEmittedData => {
        const ev = wrapper.emitted(eventName) as PaneEmittedData[][];
        const data = ev[0][0];
        expect(data).toBe(direction);
        return data;
      };

      test.each(Object.values(PaneDirections))(
        `should emit the "${Emits.DRAG_START} -> ${Emits.DRAG_MOVE} -> ${Emits.DRAG_END}" event, which takes a direction argument "%s"`,
        async (direction) => {
          const wrapper = mount(Vue3BbResize, {
            props: {
              directions: direction,
            },
          });
          await wrapper.setProps({ directions: direction });

          const paneWrapper = wrapper.findComponent(Vue3BbResizePane);
          expect(paneWrapper.exists()).toBeTruthy();

          const pane = paneWrapper.find(
            `[data-testid="${DataTestIds.SPLITTER}"]`,
          );

          let events: PaneEmittedData[] = [];

          const makeTrigger = async (
            event: string,
            eventData: { clientX: number; clientY: number },
            emittedEventName: string,
          ) => {
            await pane.trigger(event, eventData);
            paneWrapper.emitted(emittedEventName) as PaneEmittedData[][];
            events.push(checkData(emittedEventName, wrapper, direction));
          };

          await makeTrigger(
            "pointerdown",
            { clientX: 1, clientY: 1 },
            Emits.DRAG_START,
          );
          await makeTrigger(
            "pointermove",
            { clientX: 2, clientY: 2 },
            Emits.DRAG_MOVE,
          );
          await makeTrigger(
            "pointerup",
            { clientX: 3, clientY: 3 },
            Emits.DRAG_END,
          );

          expect(events.length).toBe(3);
          expect(events).toMatchSnapshot();
        },
      );

      test.each([
        [
          {
            alias: PaneDirectionAliases.HORIZONTAL,
            directions: [PaneDirections.RIGHT, PaneDirections.LEFT],
          },
        ],
        [
          {
            alias: PaneDirectionAliases.VERTICAL,
            directions: [PaneDirections.TOP, PaneDirections.BOTTOM],
          },
        ],
      ])(
        `check aliases: should emit "${Emits.DRAG_START}" with direction argument "%o"`,
        async ({ alias, directions }) => {
          const wrapper = mount(Vue3BbResize, {
            props: {
              directions: alias,
            },
          });

          const paneEls = wrapper.findAll(
            `[data-testid="${DataTestIds.SPLITTER}"]`,
          );

          directions.map(async (_d, idx) => {
            await paneEls[idx].trigger("pointerdown", {
              clientX: 1,
              clientY: 1,
            });
          });

          const ev = wrapper.emitted(Emits.DRAG_START) as PaneEmittedData[][];

          let _directions = [];
          for (const direction of directions) _directions.push(direction);

          expect(ev.length).toBe(directions.length);
          expect(_directions).toMatchObject(directions);
          expect(_directions).toMatchSnapshot();
        },
      );
    });
  });

  describe("styles", () => {
    describe("container", () => {
      test.each([{ background: "red" }])(
        "should apply custom styles (%s)",
        (container: HTMLAttributes["style"]) => {
          const wrapper = mount(Vue3BbResize, {
            props: {
              width: 320,
              height: 400,
              styles: {
                container: { style: container },
              },
            },
          });

          const containerEl = wrapper.find(
            `[data-testid="${DataTestIds.ROOT}"]`,
          );
          const styles = containerEl.attributes("style");

          // expect(styles).toContain("background: red");
          expect(styles).toMatchSnapshot();
        },
      );

      test.each([{ background: "blue", display: "flex" }])(
        "should apply inline styles (%s)",
        (container: HTMLAttributes["style"]) => {
          const wrapper = shallowMount(Vue3BbResize, {
            props: {
              style: container,
            },
          });

          const containerEl = wrapper.find(
            `[data-testid="${DataTestIds.ROOT}"]`,
          );
          const styles = containerEl.attributes("style");

          expect(styles).toContain("display: flex");
          expect(styles).toContain("background: blue");
          expect(styles).toMatchSnapshot();
        },
      );
    });
  });
});
