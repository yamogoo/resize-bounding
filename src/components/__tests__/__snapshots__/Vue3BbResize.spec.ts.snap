// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vue3BbResize > classes > prefix > should have custom prefix (my-prefix) 1`] = `
[
  "my-prefix",
]
`;

exports[`Vue3BbResize > classes > prefix > should have default prefix (%s) 1`] = `
[
  "bb-resize",
]
`;

exports[`Vue3BbResize > classes > should have "--l" class 1`] = `true`;

exports[`Vue3BbResize > components > Should not display any selected borders when set to "disabled" (--b) 1`] = `0`;

exports[`Vue3BbResize > components > Should not display any selected borders when set to "disabled" (--h) 1`] = `0`;

exports[`Vue3BbResize > components > Should not display any selected borders when set to "disabled" (--l) 1`] = `0`;

exports[`Vue3BbResize > components > Should not display any selected borders when set to "disabled" (--r) 1`] = `0`;

exports[`Vue3BbResize > components > Should not display any selected borders when set to "disabled" (--t) 1`] = `0`;

exports[`Vue3BbResize > components > Should not display any selected borders when set to "disabled" (--v) 1`] = `0`;

exports[`Vue3BbResize > components > should not render any pane element 1`] = `0`;

exports[`Vue3BbResize > components > should not render any pane element 2`] = `0`;

exports[`Vue3BbResize > components > should not render any pane element 3`] = `0`;

exports[`Vue3BbResize > components > should render bl panes (panes computed value) 1`] = `2`;

exports[`Vue3BbResize > components > should render bltr panes (panes computed value) 1`] = `4`;

exports[`Vue3BbResize > components > should render ltrb panes (panes computed value) 1`] = `4`;

exports[`Vue3BbResize > components > should render rl panes (panes computed value) 1`] = `2`;

exports[`Vue3BbResize > components > should render tr panes (panes computed value) 1`] = `2`;

exports[`Vue3BbResize > events > drag > check aliases: should emit "drag:start" with direction argument "{ alias: 'h', directions: [ 'r', 'l' ] }" 1`] = `
[
  "r",
  "l",
]
`;

exports[`Vue3BbResize > events > drag > check aliases: should emit "drag:start" with direction argument "{ alias: 'v', directions: [ 't', 'b' ] }" 1`] = `
[
  "t",
  "b",
]
`;

exports[`Vue3BbResize > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "b" 1`] = `
[
  "b",
  "b",
  "b",
]
`;

exports[`Vue3BbResize > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "l" 1`] = `
[
  "l",
  "l",
  "l",
]
`;

exports[`Vue3BbResize > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "r" 1`] = `
[
  "r",
  "r",
  "r",
]
`;

exports[`Vue3BbResize > events > drag > should emit the "drag:start -> drag:move -> drag:end" event, which takes a direction argument "t" 1`] = `
[
  "t",
  "t",
  "t",
]
`;

exports[`Vue3BbResize > slots > should render default slot 1`] = `
"<div data-testid="bb-resize-container" class="bb-resize" style="position: relative;">
  <p>Inner Container</p>
  <div class="bb-resize__pane --l" data-testid="bb-resize-pane">
    <div data-testid="bb-resize-splitter" class="bb-resize__splitter hide">
      <div data-testid="bb-resize-knob" class="bb-resize__knob" style="transform: rotate(0deg);"></div>
    </div>
  </div>
  <div class="bb-resize__pane --r" data-testid="bb-resize-pane">
    <div data-testid="bb-resize-splitter" class="bb-resize__splitter hide">
      <div data-testid="bb-resize-knob" class="bb-resize__knob" style="transform: rotate(0deg);"></div>
    </div>
  </div>
  <div class="bb-resize__pane --b" data-testid="bb-resize-pane">
    <div data-testid="bb-resize-splitter" class="bb-resize__splitter hide">
      <div data-testid="bb-resize-knob" class="bb-resize__knob" style="transform: rotate(90deg);"></div>
    </div>
  </div>
  <div class="bb-resize__pane --t" data-testid="bb-resize-pane">
    <div data-testid="bb-resize-splitter" class="bb-resize__splitter hide">
      <div data-testid="bb-resize-knob" class="bb-resize__knob" style="transform: rotate(90deg);"></div>
    </div>
  </div>
</div>"
`;

exports[`Vue3BbResize > styles > container > should apply custom styles ({ background: 'red' }) 1`] = `"position: relative; background: red;"`;

exports[`Vue3BbResize > styles > container > should apply inline styles ({ background: 'blue', display: 'flex' }) 1`] = `"background: blue; display: flex; position: relative;"`;
